src/cdomains/arrayDomain.ml

```diff
@@ -941,7 +994,746 @@ struct
   let to_yojson (x, y) = `Assoc [ (Base.name (), Base.to_yojson x); ("length", Idx.to_yojson y) ]
 end
 
-module AttributeConfiguredArrayDomain(Val: LatticeWithSmartOps) (Idx:IntDomain.Z):S with type value = Val.t and type idx = Idx.t =
+module NullByte (Val: LatticeWithNull) (Idx: IntDomain.Z): Str with type value = Val.t and type idx = Idx.t =
+struct
+  module MustNulls = SetDomain.Reverse (SetDomain.ToppedSet (IntDomain.BigInt) (struct let topname = "All Null" end))
+  module MayNulls = SetDomain.ToppedSet (IntDomain.BigInt) (struct let topname = "All Null" end)
+  (* (Must Null Set, May Null Set, Array Size) *)
+  include Lattice.Prod3 (MustNulls) (MayNulls) (Idx)
+
+  let name () = "arrays containing null bytes"
+  type idx = Idx.t
+  type value = Val.t
+
+  type ret = Null | NotNull | Top
+
+  (* helper: returns Idx.maximal except for Overflows that are mapped to None *)
+  let idx_maximal i = match Idx.maximal i with
+    | Some i -> (try Some (Z.of_int (Z.to_int i)) with Z.Overflow -> None)
+    | None -> None
+
+  let get ?(checkBounds=true) (ask: VDQ.t) (must_nulls_set, may_nulls_set, size) (e, i) =
+    let rec all_indexes_must_null i max =
+      if Z.gt i max then
+        true
+      else if MustNulls.mem i must_nulls_set then
+        all_indexes_must_null (Z.succ i) max
+      else
+        false in
```

Comment:
This looks expensive with all the `mem` queries. Maybe we can do a cardinality based thing first?

Reply:
Is this what you have in mind?

```ocaml
let all_indexes_must_null i max =
      let rec check_all_indexes i =
        if Z.gt i max then
          true
        else if MustNulls.mem i must_nulls_set then
          check_all_indexes (Z.succ i)
        else
          false in
      if Z.lt (Z.sub max i) (Z.of_int (MustNulls.cardinal must_nulls_set)) then
        false
      else
        check_all_indexes i
```

Reply:
yup